src/client/inference_session.py

```diff
@@ -154,60 +157,145 @@ def __exit__(self, *exc_details):
+                        backup_spans = self._sequence_manager.make_sequence(block_idx, recovery_end)
+                        if recovery_end is not None:
+                            # make_sequence() could return a longer sequence
+                            backup_spans[-1].end = min(backup_spans[-1].end, recovery_end)
+
+                        # The code below works even if server `server_idx` is not added yet
+                        self._exit_server_sessions(self._server_sessions[server_idx : server_idx + 1])
+                        self._chosen_spans[server_idx : server_idx + 1] = backup_spans
+                        self._server_sessions[server_idx : server_idx + 1] = self._enter_server_sessions(backup_spans)
+                        if server_idx >= len(self._server_inputs):
+                            self._server_inputs.append(None)
+                        self._server_inputs[server_idx + 1 : server_idx + 1] = [None] * (len(backup_spans) - 1)
+                        logger.debug(
+                            f"Found path from block {block_idx} to {recovery_end} via {len(backup_spans)} servers"
+                        )
+
+                    session = self._server_sessions[server_idx]
+                    span = self._chosen_spans[server_idx]
+
+                    if self._server_inputs[server_idx] is None:
+                        self._server_inputs[server_idx] = inputs
+                    elif self._server_inputs[server_idx].shape[1] == self._position:
+                        self._server_inputs[server_idx] = torch.cat(
+                            [self._server_inputs[server_idx], inputs[:, -n_input_tokens:]], dim=1
+                        )
+                    assert self._server_inputs[server_idx].shape[1] == self._position + n_input_tokens
+
+                    if recovery_end is not None and block_idx < recovery_end:
+                        inputs = self._server_inputs[server_idx]  # Pass full inputs including prefix
+                    elif block_idx == recovery_end:
+                        inputs = inputs[:, -n_input_tokens:]  # No need to pass prefix further
+
+                    outputs = session.step(inputs, prompts[span.start : span.end], **kwargs)
+                    assert outputs.shape == inputs.shape, f"expected {inputs.shape}, got {outputs.shape}"
+
+                    inputs = outputs
+                    server_idx += 1
+                    block_idx = span.end
+                    break
+                except Exception as e:
+                    delay = self._sequence_manager.min_backoff * 2**attempt_no
+                    logger.warning(
+                        f"Caught exception when running inference from block {block_idx} "
+                        f"(retry in {delay:.0f} sec): {repr(e)}"
+                    )
+                    logger.debug("See detailed traceback below:", exc_info=True)
+                    time.sleep(delay)
```

Comment:
I'd recommend that (at least) the first retry happen **without delay**

Here, if attempt_no==0, delay==min_backoff